src/core/pull/index.ts

```diff
@@ -0,0 +1,393 @@
+/*!
+ * V4Fire Core
+ * https://github.com/V4Fire/Core
+ *
+ * Released under the MIT license
+ * https://github.com/V4Fire/Core/blob/master/LICENSE
+ */
+
+/**
+ * [[include:core/pull/README.md]]
+ * @packageDocumentation
+ */
+
+import { EventEmitter2 as EventEmitter } from 'eventemitter2';
+import { resolveAfterEvents } from 'core/event';
+
+import SyncPromise from 'core/promise/sync';
+
+import { generate, serialize } from 'core/uuid';
+import { Queue } from 'core/queue';
+
+import type { PullHook, PullOptions, PullReturnType } from 'core/pull/interface';
+import { defaultValue, hashProperty, viewerCount } from 'core/pull/const';
+
+export * from 'core/pull/const';
+export * from 'core/pull/interface';
+
+/**
+ * Implementation of pull
+ * @typeparam T - pull element
+ */
+export default class Pull<T> {
+	/**
+	 * Factory from constructor argument
+	 */
+	protected objectFactory: (...args: unknown[]) => T;
+
+	/**
+	 * Data structure that contain pull's object
+	 */
+	protected readonly resourceStore: Map<string, T[]> = new Map();
+
+	/**
+	 * Data structure that contain pull's shared object
+	 */
+	protected readonly borrowedResourceStore: Map<string, T> = new Map();
+
+	/**
+	 * Hook that is activated before (free) function
+	 *
+	 * @param value - value from free(value)
+	 * @param pull - this pull
+	 * @param args - args that are given to free(value,...args)
+	 */
+	protected onFree?: PullHook<T>;
+
+	/**
+	 * Hook that is activated before `this.take` or `this.takeOrCreate`
+	 *
+	 * @param value - value that are returned from `this.take`
+	 * @param pull - this pull
+	 * @param args - args in `this.take(...args)`
+	 */
+	protected onTake?: PullHook<T>;
+
+	/**
+	 * Function that calculate hash of resource
```

Comment:
Calculates a hash of the resource

Reply:
Returns a hash value for the specified arguments